src/erleans_partitions.erl

```diff
@@ -0,0 +1,164 @@
+%%% Copyright Space-Time Insight 2017. All Rights Reserved.
+%%%
+%%% Licensed under the Apache License, Version 2.0 (the "License");
+%%% you may not use this file except in compliance with the License.
+%%% You may obtain a copy of the License at
+%%%
+%%%     http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%% Unless required by applicable law or agreed to in writing, software
+%%% distributed under the License is distributed on an "AS IS" BASIS,
+%%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%% See the License for the specific language governing permissions and
+%%% limitations under the License.
+%%%----------------------------------------------------------------------------
+
+%%% ---------------------------------------------------------------------------
+%%% @doc Each erleans node is responsible for a subset of partitions used to
+%%%      distribute responsibility for streams and reminders in the cluster.
+%%%      This server is notified by partisan when the cluster membership
+%%%      change and recalculates the range of partitions to handle.
+%%%
+%%%      Ranges are closed intervals, [Start, End]. So searching for streams
+%%%      or reminders within the interval is to include the Start and End
+%%%      partitions, i.e. Partition >=Start andalso Partition =< End.
+%%% @end
+%%% ---------------------------------------------------------------------------
+-module(erleans_partitions).
+
+-behaviour(gen_server).
+
+-export([start_link/0,
+         find_node/1,
+         add_handler/2,
+         get_range/0]).
+
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-include("erleans.hrl").
+
+-type range() :: {integer(), integer()}.
+
+-record(state, {range          :: range() | undefined,
+                num_partitions :: integer(),
+                node_ranges    :: [{range(), node()}],
+                to_notify      :: #{atom() => pid() | atom()}}).
+
+-define(CH(Item, Partitions), jch:ch(Item, Partitions)).
+
+-spec start_link() -> {ok, pid()} | {error, any()}.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec find_node(term()) -> {integer(), node()}.
+find_node(Item) ->
+    gen_server:call(?MODULE, {find_node, Item}).
+
+-spec get_range() -> range().
+get_range() ->
+    gen_server:call(?MODULE, get_range).
+
+-spec add_handler(atom(), pid() | atom()) -> ok.
+add_handler(Name, Pid) ->
+    gen_server:call(?MODULE, {add_handler, Name, Pid}).
+
+
+init([]) ->
+    %% callback for changes to cluster membership
+    Self = self(),
+    partisan_peer_service_events:add_sup_callback(fun(Membership) ->
+                                                      Self ! {update, Membership}
+                                                  end),
+    NumPartitions = erleans_config:get(num_partitions),
```

Comment:
If you just mean what happens during a change in `num_partitions` that is rolled out, instead of each node being updated in parallel, then it is simply that there will potentially be overlap in nodes handling streams causing the same stream events to be sent multiple times to subscribers.

This is one reason why we only guarantee "at least once" delivery. 

I plan to add use of backoff on # of recent changes triggered by partisan in order for the case of a rollout like that to have the re-balancing occur less because it'll wait longer and longer after a notification of cluster change before re-balancing as more come in.